src/fsharp/FSharp.Core/map.fs

```diff
@@ -64,33 +51,30 @@ namespace Microsoft.FSharp.Collections
 
         let empty = MapEmpty 
 
-        let height  = function
-          | MapEmpty -> 0
-          | MapOne _ -> 1
-          | MapNode(_,_,_,_,h) -> h
+        let size x =
+            match x with
+            | MapEmpty -> 0
+            | MapNode (_,_,_,_,h) -> h
 
         let isEmpty m = 
             match m with 
             | MapEmpty -> true
             | _ -> false
 
-        let mk l k v r = 
-            match l,r with 
-            | MapEmpty,MapEmpty -> MapOne(k,v)
-            | _ -> 
-                let hl = height l 
-                let hr = height r 
-                let m = if hl < hr then hr else hl 
-                MapNode(k,v,l,r,m+1)
+        let inline mk l k v r =
+            MapNode (k,v,l,r, size l + size r + 1)
+
+        let inline mkLeaf k v =
+            MapNode (k, v, MapEmpty, MapEmpty, 1)
 
         let rebalance t1 k v t2 =
-            let t1h = height t1 
-            let t2h = height t2 
-            if  t2h > t1h + 2 then (* right is heavier than left *)
+            let t1h = size t1 
+            let t2h = size t2 
+            if  (t2h >>> 1) > t1h then (* right is heavier than left *)
```

Comment:
You changed from -2 to /2 is this intended? Is it because we now control size instead of height? If so then the identifier should be renamed to reflect that